### ISSUE TITLE
Raycasting the terrain can be very slow (main cause of slowdown in Godot 3.0.6)

### ISSUE THREAD

**Issue by valentinAlkan** [Zylann/godot_heightmap_plugin#57]
I have a very simple scene set up with a very simple terrain, a directional light, and a camera. When collision is enabled and the scene is run or played in the editor, the scene struggles to achieve more than a few fps. If collision is disabled, the scene runs normally.

Is there some setting I'm missing? I would like to use collision with my project if possible.

**Comment by Zylann** [comment]
Collision is achieved using Bullet's heightmap shape, which is built in Godot. Whan my plugin does is only to provide it the heightmap, so I'm not sure what could affect FPS so much...
Do you have a project I can check to see if it runs slowly on my end?

**Comment by valentinAlkan** [comment]
There is a little bit of extra in this one but it shows the issue better. There is a profiler in the game itself so you can see the FPS during export.

[newerhide.zip](https://github.com/Zylann/godot_heightmap_plugin/files/2683248/newerhide.zip)

**Comment by valentinAlkan** [comment]
This is a simpler project. I still see significant changes in FPS when collision enabled, but since its a smaller collider it does not appear to happens as frequently and as severely.

The profiler is not intended to be a replacement for godot's embedded profiler (though they do report the same values from what I can tell), but mostly its there so I can see FPS profile of an exported project. It's only as accurate as Engine.get_frames_per_second() is.

[terrain_test.zip](https://github.com/Zylann/godot_heightmap_plugin/files/2683272/terrain_test.zip)

**Comment by Zylann** [comment]
About the profiler, I was talking about the Godot one :)

I'm still looking at the previous test project, and I found FPS is okay if I make the camera look down. But as soon as I place the mouse above terrain pixels, performance goes down... interesting!

**Comment by valentinAlkan** [comment]
Ah, gotcha :).

yeah its weird. I hope this can be resolved quickly, Im currently trying to produce a proof-of-concept to replace an Unreal project we have at my work with godot. but it needs a terrain system and collision. 

For what its worth, which might not be much, but I have seen this problem on 3 different systems - 2 linux and one windows. So whatever it is, its consistent between windows and linux, anyway.

**Comment by Zylann** [comment]
Oh god... I reproduced it. Just the collider in an empty project, moving my mouse over the place it would be, and slows down immediately when I'm over it. Magic ✨

**Comment by Zylann** [comment]
I investigated what's going on in the engine (see later comments on the issue linked above), and it looks like only raycasting causes this. You might be able to workaround the slowness in Godot 3.1 by turning off picking (which should be off by default, otherwise use `PhysicsServer.body_set_ray_pickable(_body_rid, false)` in the collider script).

**Comment by Zylann** [comment]
I confirm this is extremely bad on lower-spec machines, so at this point I would start migrating the plugin to Godot 3.1 because at least it allows to bypass the picking issue.

I just posted 0.10 for release on the assetlib, and it will probably be the last on Godot 3.0.6 for this reason.

Also, to adress the underlying problem, I opened an issue on Bullet's repo: https://github.com/bulletphysics/bullet3/issues/2032 on which I might work on when I get some time.

**Comment by pgruenbacher** [comment]
bullet terrain physics is pretty slow for ray-casting from what I know of since it just basically interrogates all the triangles that lie under the raycast aabb.

**Comment by Zylann** [comment]
@pgruenbacher I've been working on that https://github.com/Zylann/godot/commit/e780160d99d83a34a31f66470f5f4d832fcff9db

**Comment by pgruenbacher** [comment]
ah so you're doing bridge triangle traversal. 
Personally, since I only care about raycasting and not collisions involving terrain, what I do is simply increment the ray by it's horizontal component, check to see if it's under the terrain, and if so, interpolate the height.  It's not as accurate as triangles, but for terrain picking and for raycast physics it works well enough for my needs.

**Comment by pgruenbacher** [comment]
```
bool TerrainManager::testRay(const Physics::Ray& ray, prop_t& dist) const {
  // ray tests just need to be approximate for agent projectiles.
  if (!valid()) return false;
  // ray cannot start below terrain height.
  auto length = ray.length;
  assert(ray.normal != Vector3(0,0,0));
  // std::cout << "RAY x0 " << ray.x0 << std::endl;
  assert(interpolateHeight(ray.x0[0], ray.x0[2]) < ray.x0[1]);
  auto pos = ray.x0;

  auto finalPos = ray.finalPos;

  auto vec2d = Menge::Math::Vector2(ray.normal[0], ray.normal[2]);
  real_pos_t horiz_length = vec2d.length();
  // if horiz very small, it's vertical...
  if (horiz_length < 0.0001) {
      // the ray is practically vertical up, so it's not gonna intersect...
    if (ray.normal[1] > 0) return false;

    // practically vertical, so just interpolate final pos height and compare.
    auto terrainH = interpolateHeight(ray.finalPos[0], ray.finalPos[2]);
    if (terrainH > ray.finalPos[1]) {
      // dist = ray.x0[1] - terrainH;
      dist = length;
      return true;
    }
    return false;
  }


  auto step = fraction(_cellS

  dist = 0.f;
  while (dist < length) {
      pos[0] += (ray.normal[0] * step);
      pos[1] += (ray.normal[1] * step);
      pos[2] += (ray.normal[2] * step);
      dist += step;
      if (interpolateHeight(pos[0], pos[2]) > pos[1]) {
        return true;
      }
  }

  return false;

}
```

**Comment by pgruenbacher** [comment]
but I wouldn't do what I do since my implementation is very sloppy, I just needed much much higher performance for the game runtime (thousands of projectiles, visibility raycasts)

**Comment by Zylann** [comment]
@pgruenbacher yeah I would prefer PR a precise general-case raycast to Bullet, but for a faster lower-precision that could be added to the plugin, because I need a physics-less raycast to paint stuff on it. However it's in GDScript^^

**Comment by Zylann** [comment]
I'll close this since it's been fixed starting from Godot 3.2.
